Batteries/Data/List/Lemmas.lean

```diff
@@ -482,6 +491,34 @@ theorem Sublist.erase_diff_erase_sublist {a : α} :
 
 end Diff
 
+/-! ### prefix, suffix, infix -/
+
+theorem singleton_prefix_cons (a) : [a] <+: a :: l :=
+  (prefix_cons_inj a).mpr nil_prefix
+
+theorem ne_nil_of_not_prefix (h : ¬l₁ <+: l₂) : l₁ ≠ [] := by
+  intro heq
+  simp [heq, nil_prefix] at h
+
+theorem not_prefix_and_not_prefix_symm_iff_exists [BEq α] [LawfulBEq α] [DecidableEq α]
+    {l₁ l₂ : List α} : ¬l₁ <+: l₂ ∧ ¬l₂ <+: l₁ ↔ ∃ c₁ c₂ pre suf₁ suf₂, c₁ ≠ c₂ ∧
+      l₁ = pre ++ c₁ :: suf₁ ∧ l₂ = pre ++ c₂ :: suf₂ := by
```

Comment:
Since all the `∃` parts are unique, we might as well have a function that computes the pieces and then prove the specs of that function.

Reply:
Namely:
```
def matchPrefix (l₁ l₂ : List α) : List α × List α × List α := ...

theorem match_prefix_spec : matchPrefix l₁ l₂ = (p, t₁, t₂) ↔ t₁.head? ≠ t₂.head? ∧ l₁ = p ++ t₁ ∧ l₂ = p ++ t₂ := ...
```

Reply:
`t₁.head? ≠ t₂.head? ` is wrong there, it fails when both lists are equal and the tails are empty

Reply:
Sorry, I was typing too quick. The lemma assumes that neither list is a prefix of each other. This is a better version:
```lean
theorem matchPrefix_spec : matchPrefix l₁ l₂ = (p, t₁, t₂) ↔ l₁ = l₂ ∨ t₁.head? ≠ t₂.head? ∧ l₁ = p ++ t₁ ∧ l₂ = p ++ t₂ := ...
```

Reply:
I renamed and restated the theorem as follows:

```lean
List.theorem not_prefix_and_not_prefix_symm_iff {l₁ l₂ p t₁ t₂ : List α} (hp : commonPrefix l₁ l₂ = p)
    (h₁ : l₁ = p ++ t₁) (h₂ : l₂ = p ++ t₂) : ¬l₁ <+: l₂ ∧ ¬l₂ <+: l₁ ↔ t₁ ≠ [] ∧ t₂ ≠ [] ∧
      t₁.head? ≠ t₂.head? := by
  sorry
```